src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,150 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    public static class Node<T> {
+        private Node<T> next;
+        private Node<T> prev;
+        private T value;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.next = next;
+            this.prev = prev;
+            this.value = value;
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            head = new Node<>(null, value, null);
+            tail = head;
+        } else {
+            Node<T> nodeToAdd = new Node<>(tail, value, null);
+            tail.next = nodeToAdd;
+            tail = nodeToAdd;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIfIndexOutOfBonds(index, size);
+        if ((index == 0 && head == null) || index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            head = new Node<>(null, value, head);
+        } else {
+            Node<T> previous = getNodeByIndex(index - 1);
+            Node<T> nodeToAdd = new Node<>(previous, value, previous.next);
+            previous.next.prev = nodeToAdd;
+            previous.next = nodeToAdd;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIfIndexOutOfBonds(index, size - 1);
+        Node<T> current = getNodeByIndex(index);
+        return current.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIfIndexOutOfBonds(index, size - 1);
+        Node<T> current = getNodeByIndex(index);
+        T valueBeforeSet = current.value;
+        current.value = value;
+        return valueBeforeSet;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIfIndexOutOfBonds(index, size - 1);
+        T removedValue;
+        if (index == 0) {
+            removedValue = head.value;
+            unlink(head);
+        } else {
+            Node<T> previous = getNodeByIndex(index - 1);
+            removedValue = previous.next.value;
+            unlink(previous.next);
+        }
```

Comment:
The `size` of the list should be decremented after removing an element. Add `size--;` before returning `removedValue`.